services/background-worker/src/Wire/BackendNotificationPusher.hs

```diff
@@ -78,32 +82,99 @@ pushNotification runningFlag targetDomain (msg, envelope) = do
+            )
+            >>= \case
+              Left e -> do
+                Log.err $
+                  Log.msg (Log.val "Failed to get supported API versions, the notification will be ignored")
+                    . Log.field "domain" (domainText targetDomain)
+                    . Log.field "error" (displayException e)
+                throwM e
+              Right vi -> pure . Set.fromList . vinfoSupported $ vi
 
-        -- FUTUREWORK: This rejects the message without any requeueing. This is
-        -- dangerous as it could happen that a new type of notification is
-        -- introduced and an old instance of this worker is running, in which case
-        -- the notification will just get dropped. On the other hand not dropping
-        -- this message blocks the whole queue. Perhaps there is a better way to
-        -- deal with this.
-        lift $ reject envelope False
-      Right notif -> do
-        ceFederator <- asks (.federatorInternal)
-        ceHttp2Manager <- asks http2Manager
-        let ceOriginDomain = notif.ownDomain
-            ceTargetDomain = targetDomain
-            ceOriginRequestId = fromMaybe (RequestId "N/A") notif.requestId
-            fcEnv = FederatorClientEnv {..}
-        liftIO $ either throwM pure =<< sendNotification fcEnv notif.targetComponent notif.path notif.body
-        lift $ ack envelope
-        metrics <- asks backendNotificationMetrics
-        withLabel metrics.pushedCounter (domainText targetDomain) incCounter
-        withLabel metrics.stuckQueuesGauge (domainText targetDomain) (flip setGauge 0)
+        -- compute the best usable version in a notification
+        let bestVersion = bodyVersions >=> flip latestCommonVersion remoteVersions
+        case pairedMaximumOn bestVersion (toList (notifications bundle)) of
+          (_, Nothing) ->
+            Log.fatal $
+              Log.msg (Log.val "No federation API version in common, the notification will be ignored")
+                . Log.field "domain" (domainText targetDomain)
+          (notif, cveVersion) -> do
+            ceFederator <- asks (.federatorInternal)
+            ceHttp2Manager <- asks http2Manager
+            let ceOriginDomain = notif.ownDomain
+                ceTargetDomain = targetDomain
+                ceOriginRequestId = fromMaybe (RequestId "N/A") notif.requestId
+                cveEnv = FederatorClientEnv {..}
+                fcEnv = FederatorClientVersionedEnv {..}
+            liftIO $ either throwM pure =<< sendNotification fcEnv notif.targetComponent notif.path notif.body
+            lift $ ack envelope
+            metrics <- asks backendNotificationMetrics
+            withLabel metrics.pushedCounter (domainText targetDomain) incCounter
+            withLabel metrics.stuckQueuesGauge (domainText targetDomain) (flip setGauge 0)
+
+-- | Find the pair that maximises b.
+pairedMaximumOn :: Ord b => (a -> b) -> [a] -> (a, b)
+pairedMaximumOn f = maximumBy (compare `on` snd) . map (id &&& f)
```

Comment:
I don't think a test would have a lot of value here. The types are general enough that there is essentially only one implementation. Well, two, since there is a `minimum` version, but hopefully it's pretty clear which one this is. Not sure how a test would make that clearer.